lrauv_description/scripts/description_generator.py

```diff
@@ -45,92 +45,142 @@ def write_float_array(pose):
+                dz = float(mass.text) / (dx * dy * fluid_density)
+                size.text = write_float_array([dx, dy, dz])
+                continue
+            else:
+                assert False, "Detected unknown collision"
+
+        # Only these links get here. Checking to be safe because the script
+        # isn't generic enough yet.
+        link_name = link_tag.get("name")
+        assert link_name == "base_link" or link_name == "battery" or link_name == "drop_weight", link_name
+
+        # Get various tags used to calculate moments
+        inertial_tag = link_tag.find(".inertial")
+        assert inertial_tag is not None, "Missing inertial"
+
+        mass_tag = link_tag.find(".inertial/mass")
+        assert mass_tag is not None, "Missing mass"
+
+        link_pose_tag = link_tag.find(".pose")
+        inertial_pose_tag = link_tag.find(".inertial/pose")
+
+        # Hold the base link tags to fill below
+        if link_name == "base_link":
+            base_link_mass_tag = mass_tag
+            base_link_inertial_pose_tag = inertial_pose_tag
+            continue
+
+        # Get the CoM pose w.r.t. the model origin for each link
+        # Currently only support links with either <link><pose> or
+        # <inertial><pose>, but not both, and without rotation.
+        if inertial_pose_tag is not None and link_pose_tag is None:
+            center_of_mass = [float(token) for token in inertial_pose_tag.text.split()]
+        elif inertial_pose_tag is None and link_pose_tag is not None:
+            center_of_mass = [float(token) for token in link_pose_tag.text.split()]
+        else:
+            assert False, "Links with both <link><pose> and <inertial><pose> aren't supported yet."
+
+        assert len(center_of_mass) == 6
+
+        # Rotation isn't properly handled yet
+        assert center_of_mass[3] == 0.0 and \
+               center_of_mass[4] == 0.0 and \
+               center_of_mass[5] == 0.0, "Link rotation isn't properly handled yet."
+
+        # Moments about the X axis
+        x_moments.append((float(mass_tag.text), center_of_mass[0]))
+
+    # Moments and mass for non-neutral links other than the base link (i.e. components)
+    component_x_moments =  sum([mass * com for mass, com in x_moments])
+    component_mass = sum([mass for mass, _ in x_moments])
+    assert component_mass < total_mass, "Component mass is larger than total mass"
+
+    # Calculate the base link's inertial pose and mass
+    remaining_mass = total_mass - component_mass
+    base_link_inertial_pose_x = - component_x_moments / remaining_mass
+
+    base_link_mass_tag.text = str(remaining_mass)
+    base_link_inertial_pose_tag.text = write_float_array([base_link_inertial_pose_x, 0, 0, 0, 0, 0])
+
+    # Calculate base link's volume size and position so that it cancels out the
+    # mass of the base link, as well as the components
+    base_link_dx = 2.0
+    base_link_dy = 0.4
+    base_link_dz = total_mass / (2 * 0.4 * fluid_density)
```

Comment:
```suggestion
    base_link_dz = total_mass / (base_link_dx *  base_link_dy * fluid_density)
```
Knit: lets also move `base_link_dx` and `base_link_dy` up to the top maybe so they can be parameters?

Reply:
Nice catch, fixed :+1: